components/raftstore/src/store/fsm/peer.rs

```diff
@@ -4749,27 +4791,132 @@ where
             );
+
+            // test purpose
+            #[cfg(any(test, feature = "testexport"))]
+            {
+                let default_buckets = BucketStat::default();
+                test_only_callback(
+                    _cb,
+                    self.fsm
+                        .peer
+                        .region_buckets
+                        .as_ref()
+                        .unwrap_or(&default_buckets)
+                        .meta
+                        .clone(),
+                );
+            }
             return;
         }
-        let mut meta = BucketMeta {
-            region_id: self.fsm.region_id(),
-            region_epoch,
-            version: timespec_to_ns(monotonic_raw_now()),
-            keys: bucket_keys,
-        };
-        meta.keys.insert(0, region.get_start_key().to_vec());
-        meta.keys.push(region.get_end_key().to_vec());
-        let stats = new_bucket_stats(&meta);
-        let meta = Arc::new(meta);
-        let region_buckets = BucketStat::new(meta.clone(), stats);
-        if self.fsm.peer.region_buckets.is_none() {
-            self.register_report_region_buckets_tick();
+
+        let current_version = self
+            .fsm
+            .peer
+            .region_buckets
+            .as_ref()
+            .map(|b| b.meta.version)
+            .unwrap_or_default();
+        let mut region_buckets: BucketStat;
+        if let Some(bucket_ranges) = bucket_ranges {
+            assert_eq!(buckets.len(), bucket_ranges.len());
+            let mut i = 0;
+            let mut j = 0;
+            region_buckets = self.fsm.peer.region_buckets.as_ref().unwrap().clone();
+            let mut meta = (*region_buckets.meta).clone();
+            for mut bucket in buckets {
+                while i < meta.keys.len() && meta.keys[i] != bucket_ranges[j].0 {
+                    i += 1;
+                }
+                assert!(i != meta.keys.len());
+                // the bucket size is small and does not have split keys,
+                // then it should be merged with its left neighbor
+                let region_bucket_merge_size =
+                    self.ctx.coprocessor_host.cfg.region_bucket_merge_size_ratio
+                        * (self.ctx.coprocessor_host.cfg.region_bucket_size.0 as f64);
+                if bucket.keys.is_empty() && bucket.size <= (region_bucket_merge_size as u64) {
+                    meta.sizes[i] = bucket.size;
+                    // i is not the last entry (which is end key)
+                    assert!(i < meta.keys.len() - 1);
+                    // the region has more than one bucket
+                    // and the left neighbor + current bucket size is not very big
+                    if meta.keys.len() > 2
+                        && i != 0
+                        && meta.sizes[i - 1] + bucket.size
+                            < self.ctx.coprocessor_host.cfg.region_bucket_size.0 * 2
```

Comment:
Why *2 here? Is it possible the bucket is splitted soon?

Reply:
it's not possible because when do split, we only allow bucket's size >= 2* region_bucket_size. 
The idea here is to reduce small size bucket as much as possible.